File: 백준/Platinum/11266. 단절점/단절점.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;
import java.util.TreeSet;

class Main {
    static StringTokenizer st;
    static List<Integer>[] edges;
    static TreeSet<Integer> res = new TreeSet<>();
    static int[] ids;
    static int V, E, id = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        V = nextInt();
        E = nextInt();

        ids = new int[V + 1];
        edges = new List[V + 1];

        for (int u = 1; u <= V; u++) {
            edges[u] = new ArrayList<>();
        }

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int u = nextInt();
            int v = nextInt();

            edges[u].add(v);
            edges[v].add(u);
        }

        for (int u = 1; u <= V; u++) {
            if (ids[u] == 0) {
                travel(u, u);
            }
        }

        StringBuilder sb = new StringBuilder();

        sb.append(res.size()).append('\n');

        for (int u : res) {
            sb.append(u).append(' ');
        }

        System.out.println(sb);
        br.close();
    }

    static int travel(int p, int u) {
        ids[u] = ++id;

        int minId = id;
        int count = 0;

        for (int v : edges[u]) {
            if (v == p) continue;

            if (ids[v] > 0) {
                minId = Math.min(minId, ids[v]);
                continue;
            }

            int cid = travel(u, v);

            if (u != p && ids[u] <= cid) {
                res.add(u);
            }

            count++;
            minId = Math.min(minId, cid);
        }

        if (u == p && count > 1) {
            res.add(u);
        }

        return minId;
    }

    static int nextInt() {
        return Integer.parseInt(st.nextToken());
    }
}

File: 백준/Platinum/11266. 단절점/단절점.kt
import java.io.StreamTokenizer

fun main() = with(StreamTokenizer(System.`in`.bufferedReader())) {
    val readInt = {
        nextToken()
        nval.toInt()
    }
    val n = readInt()
    val e = readInt()
    val edges = Array(n + 1) { ArrayList<Int>() }

    repeat(e) {
        val u = readInt()
        val v = readInt()

        edges[u] += v
        edges[v] += u
    }

    val spfPoints = sortedSetOf<Int>()
    val ids = IntArray(n + 1)
    var id = 0

    fun travel(u: Int, p: Int): Int {
        ids[u] = ++id

        if (edges[u].isEmpty()) return id

        var minId = id
        var travels = 0

        for (v in edges[u]) {
            if (ids[v] == 0) {
                minId = minOf(minId, travel(v, u))
                travels++
            }
            else if (v != p) {
                minId = minOf(minId, ids[v])
            }
        }

        if (minId >= ids[p]) spfPoints += p

        if (u == p && travels == 1) spfPoints -= u

        return minId
    }

    for (u in 1..n) {
        if (ids[u] == 0) travel(u, u)
    }

    with(System.out.bufferedWriter()) {
        append("${spfPoints.size}\n")
        spfPoints.joinTo(this, " ")
        flush()
        close()
    }
}